snapshots/rawblock/rawblock.go

```diff
@@ -0,0 +1,570 @@
+		}
+	}()
+
+	err = t.Commit()
+	t = nil
+	return err
+}
+
+// Remove the committed or active snapshot by the provided key.
+//
+// All resources associated with the key will be removed.
+//
+// If the snapshot is a parent of another snapshot, its children must be
+// removed before proceeding.
+func (o *snapshotter) Remove(ctx context.Context, key string) (err error) {
+	ctx, t, err := o.ms.TransactionContext(ctx, true)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil && t != nil {
+			if rerr := t.Rollback(); rerr != nil {
+				log.G(ctx).WithError(rerr).Warn("failed to rollback transaction")
+			}
+		}
+	}()
+
+	id, kind, err := storage.Remove(ctx, key)
+	if err != nil {
+		return errors.Wrap(err, "failed to remove")
+	}
+
+	path := o.getSnapshotPath(id, kind)
+	renamed := o.getSnapshotPath("rm-"+id, kind)
+	if err := os.Rename(path, renamed); err != nil {
+		if !os.IsNotExist(err) {
+			return errors.Wrap(err, "failed to rename")
+		}
+		renamed = ""
+	}
+
+	err = t.Commit()
+	t = nil
+	if err != nil {
+		if renamed != "" {
+			if err1 := os.Rename(renamed, path); err1 != nil {
+				// May cause inconsistent data on disk
+				log.G(ctx).WithError(err1).WithField("path", renamed).Errorf("failed to rename after failed commit")
+			}
+		}
+		return errors.Wrap(err, "failed to commit")
+	}
+	if renamed != "" {
+		if err := os.RemoveAll(renamed); err != nil {
+			// Must be cleaned up, any "rm-*" could be removed if no active transactions
+			log.G(ctx).WithError(err).WithField("path", renamed).Warnf("failed to remove snapshot file")
+		}
+	}
+
+	return nil
+}
+
+// Walk all snapshots in the snapshotter. For each snapshot in the
+// snapshotter, the function will be called.
+func (o *snapshotter) Walk(ctx context.Context, fn func(context.Context, snapshots.Info) error) error {
+	ctx, t, err := o.ms.TransactionContext(ctx, false)
+	if err != nil {
+		return err
+	}
+	defer t.Rollback()
+	return storage.WalkInfo(ctx, fn)
+}
+
+func (o *snapshotter) createBaseImage(ctx context.Context) (baseImage string, err error) {
+	var base string
+	f, err := ioutil.TempFile(filepath.Join(o.root, snapshots.KindActive.String()), "new-")
+	if err != nil {
+		return "", err
+	}
+	base = f.Name()
+	f.Close()
+	defer func() {
+		if err != nil {
+			if err1 := os.Remove(base); err1 != nil {
+				log.G(ctx).WithError(err1).Warnf("failed to remove temp file %s after base image creation failure", base)
+			}
+		}
+	}()
+	if err := os.Chmod(base, 0755); err != nil {
+		return "", err
+	}
+	if err := os.Truncate(base, int64(o.sizeMB<<20)); err != nil {
+		return "", err
+	}
+	switch o.fstype {
+	case "xfs":
+		if out, err := exec.Command("mkfs.xfs", "-f", base).CombinedOutput(); err != nil {
+			return "", errors.Errorf("Failed to create %s file system on %s: %v:%s", o.fstype, base, err, string(out))
+		}
+	case "ext4":
+		args := []string{
+			base,
+			"-F",
+			"-O",
+			"^uninit_bg",
```

Comment:
is the `^` supposed to be here?

Reply:
Yes, the idea is to disable ext4 `uninit_bg` feature so that the base image will init all block groups. Then we don't have to init them after mounting every snapshot.